services/venice-controller/src/test/java/com/linkedin/venice/controller/kafka/protocol/ProtocolCompatibilityTest.java

```diff
@@ -0,0 +1,79 @@
+package com.linkedin.venice.controller.kafka.protocol;
+
+import com.linkedin.avro.fastserde.FastSerdeCache;
+import com.linkedin.venice.schema.avro.SchemaCompatibility;
+import com.linkedin.venice.utils.TestUtils;
+import java.util.Collections;
+import java.util.Map;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import org.apache.avro.Schema;
+import org.apache.avro.SchemaValidationException;
+import org.apache.avro.SchemaValidator;
+import org.apache.avro.SchemaValidatorBuilder;
+import org.testng.Assert;
+
+
+public abstract class ProtocolCompatibilityTest {
+  protected void testProtocolCompatibility(Map<Integer, Schema> schemaMap, int latestSchemaId)
+      throws InterruptedException {
+    Assert.assertNotNull(schemaMap.containsKey(latestSchemaId), "The latest schema should exist!");
+
+    SchemaValidatorBuilder schemaValidatorBuilder = new SchemaValidatorBuilder();
+    SchemaValidator schemaValidator = schemaValidatorBuilder.canReadStrategy().validateAll();
+
+    /**
+     * Also checked the schema evolution is acceptable for fast avro.
+     */
+    ExecutorService executor = Executors.newSingleThreadExecutor();
+    try {
+      FastSerdeCache fastSerdeCache = new FastSerdeCache(executor);
+
+      Schema latestSchema = schemaMap.get(latestSchemaId);
+      schemaMap.forEach((schemaId, schema) -> {
+        if (schemaId == latestSchemaId) {
+          return;
+        }
+        SchemaCompatibility.SchemaPairCompatibility backwardCompatibility =
+            SchemaCompatibility.checkReaderWriterCompatibility(latestSchema, schema);
```

Comment:
Shouldn't we be checking both ways of reader-writer compatibility? It might be better to use `SchemaEntry.isNewSchemaCompatible`

Reply:
Did you mean by checking `schemaEntry.isNewSchemaCompatible(latestSchemaEntry, DirectionalSchemaCompatibilityType.FULL)`? This does not work for previous KME protocols. I think we can only guarantee backward compatibility by adding back `SOBR` here.

Reply:
Discussed offline. We are OK with only backward compatibility, and we plan to set up a Policy for this kind of KME upgrade.